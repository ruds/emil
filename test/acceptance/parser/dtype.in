datatype suit = Spades | Hearts | Diamonds | Clubs;
datatype 'a tree = Empty | Node of 'a * 'a forest
     and 'a forest = Nil | Cons of 'a tree * 'a forest;
datatype arith_expr = Number of bigint
                    | arith_expr ~+ arith_expr
                    | arith_expr ~- arith_expr
                    | ~- arith_expr;
datatype arith_expr = Number of bigint
                    | (arith_expr) ~+ arith_expr
                    | (~-) of arith_expr * arith_expr
                    | (prefix ~-) of arith_expr;
datatype formula = Prop of str
                 | ¬ formula
                 | formula ∧ formula
                 | formula ∨ formula
                 | formula → formula
                 | (⊥);
datatype ('a, 'b) pair = Pair of 'a * 'b;
datatype pair = Pair of 'a * 'b;
datatype 'a pair = Pair of 'a * 'b;
datatype 'a, 'b pair = Pair of 'a * 'b;
datatype pair = pair of int * string;
datatype pair = pair of int * string
     and pair = Jacks | Queens | Kings | Aces;
datatype foo = Foo | Bar | Bar of int;
datatype foo = Foo | Bar and bar = Bar | Baz;
datatype it = Foo;
datatype foo = it;
datatype ref = Foo;
datatype foo = ref;
datatype true = Foo;
datatype foo = true;
datatype false = Foo;
datatype foo = false;
datatype nil = Foo;
datatype foo = nil;
datatype Foo = (::);
datatype Foo = int :: int;
datatype Foo = (::) of (int * int);
