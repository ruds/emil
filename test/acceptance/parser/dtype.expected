0001 (DeclTopDecl
         :decl
         (DtypeDecl
             :bindings
             ((DtypeBind
                 :identifier
                 "73756974"
                 :types
                 ()
                 :constructors
                 (
                     (ConBind
                         :identifier
                         "537061646573"
                         :param
                         nullptr
                         :is_op
                         false
                         :is_prefix_op
                         false)
                     (ConBind
                         :identifier
                         "486561727473"
                         :param
                         nullptr
                         :is_op
                         false
                         :is_prefix_op
                         false)
                     (ConBind
                         :identifier
                         "4469616D6F6E6473"
                         :param
                         nullptr
                         :is_op
                         false
                         :is_prefix_op
                         false)
                     (ConBind
                         :identifier
                         "436C756273"
                         :param
                         nullptr
                         :is_op
                         false
                         :is_prefix_op
                         false)
                 )))))
0002 (DeclTopDecl
         :decl
         (DtypeDecl
             :bindings
             (
                 (DtypeBind
                     :identifier
                     "74726565"
                     :types
                     ("2761")
                     :constructors
                     (
                         (ConBind
                             :identifier
                             "456D707479"
                             :param
                             nullptr
                             :is_op
                             false
                             :is_prefix_op
                             false)
                         (ConBind
                             :identifier
                             "4E6F6465"
                             :param
                             (TupleTypeExpr
                                 :types
                                 (
                                     (VarTypeExpr
                                         :id
                                         "2761")
                                     (TyconTypeExpr
                                         :qualifiers
                                         ()
                                         :identifier
                                         "666F72657374"
                                         :types
                                         ((VarTypeExpr
                                             :id
                                             "2761")))
                                 ))
                             :is_op
                             false
                             :is_prefix_op
                             false)
                     ))
                 (DtypeBind
                     :identifier
                     "666F72657374"
                     :types
                     ("2761")
                     :constructors
                     (
                         (ConBind
                             :identifier
                             "4E696C"
                             :param
                             nullptr
                             :is_op
                             false
                             :is_prefix_op
                             false)
                         (ConBind
                             :identifier
                             "436F6E73"
                             :param
                             (TupleTypeExpr
                                 :types
                                 (
                                     (TyconTypeExpr
                                         :qualifiers
                                         ()
                                         :identifier
                                         "74726565"
                                         :types
                                         ((VarTypeExpr
                                             :id
                                             "2761")))
                                     (TyconTypeExpr
                                         :qualifiers
                                         ()
                                         :identifier
                                         "666F72657374"
                                         :types
                                         ((VarTypeExpr
                                             :id
                                             "2761")))
                                 ))
                             :is_op
                             false
                             :is_prefix_op
                             false)
                     ))
             )))
0004 (DeclTopDecl
         :decl
         (DtypeDecl
             :bindings
             ((DtypeBind
                 :identifier
                 "61726974685F65787072"
                 :types
                 ()
                 :constructors
                 (
                     (ConBind
                         :identifier
                         "4E756D626572"
                         :param
                         (TyconTypeExpr
                             :qualifiers
                             ()
                             :identifier
                             "626967696E74"
                             :types
                             ())
                         :is_op
                         false
                         :is_prefix_op
                         false)
                     (ConBind
                         :identifier
                         "7E2B"
                         :param
                         (TupleTypeExpr
                             :types
                             (
                                 (TyconTypeExpr
                                     :qualifiers
                                     ()
                                     :identifier
                                     "61726974685F65787072"
                                     :types
                                     ())
                                 (TyconTypeExpr
                                     :qualifiers
                                     ()
                                     :identifier
                                     "61726974685F65787072"
                                     :types
                                     ())
                             ))
                         :is_op
                         true
                         :is_prefix_op
                         false)
                     (ConBind
                         :identifier
                         "7E2D"
                         :param
                         (TupleTypeExpr
                             :types
                             (
                                 (TyconTypeExpr
                                     :qualifiers
                                     ()
                                     :identifier
                                     "61726974685F65787072"
                                     :types
                                     ())
                                 (TyconTypeExpr
                                     :qualifiers
                                     ()
                                     :identifier
                                     "61726974685F65787072"
                                     :types
                                     ())
                             ))
                         :is_op
                         true
                         :is_prefix_op
                         false)
                     (ConBind
                         :identifier
                         "7E2D"
                         :param
                         (TyconTypeExpr
                             :qualifiers
                             ()
                             :identifier
                             "61726974685F65787072"
                             :types
                             ())
                         :is_op
                         true
                         :is_prefix_op
                         true)
                 )))))
0008 (DeclTopDecl
         :decl
         (DtypeDecl
             :bindings
             ((DtypeBind
                 :identifier
                 "61726974685F65787072"
                 :types
                 ()
                 :constructors
                 (
                     (ConBind
                         :identifier
                         "4E756D626572"
                         :param
                         (TyconTypeExpr
                             :qualifiers
                             ()
                             :identifier
                             "626967696E74"
                             :types
                             ())
                         :is_op
                         false
                         :is_prefix_op
                         false)
                     (ConBind
                         :identifier
                         "7E2B"
                         :param
                         (TupleTypeExpr
                             :types
                             (
                                 (TyconTypeExpr
                                     :qualifiers
                                     ()
                                     :identifier
                                     "61726974685F65787072"
                                     :types
                                     ())
                                 (TyconTypeExpr
                                     :qualifiers
                                     ()
                                     :identifier
                                     "61726974685F65787072"
                                     :types
                                     ())
                             ))
                         :is_op
                         true
                         :is_prefix_op
                         false)
                     (ConBind
                         :identifier
                         "7E2D"
                         :param
                         (TupleTypeExpr
                             :types
                             (
                                 (TyconTypeExpr
                                     :qualifiers
                                     ()
                                     :identifier
                                     "61726974685F65787072"
                                     :types
                                     ())
                                 (TyconTypeExpr
                                     :qualifiers
                                     ()
                                     :identifier
                                     "61726974685F65787072"
                                     :types
                                     ())
                             ))
                         :is_op
                         true
                         :is_prefix_op
                         false)
                     (ConBind
                         :identifier
                         "7E2D"
                         :param
                         (TyconTypeExpr
                             :qualifiers
                             ()
                             :identifier
                             "61726974685F65787072"
                             :types
                             ())
                         :is_op
                         true
                         :is_prefix_op
                         true)
                 )))))
0012 (DeclTopDecl
         :decl
         (DtypeDecl
             :bindings
             ((DtypeBind
                 :identifier
                 "666F726D756C61"
                 :types
                 ()
                 :constructors
                 (
                     (ConBind
                         :identifier
                         "50726F70"
                         :param
                         (TyconTypeExpr
                             :qualifiers
                             ()
                             :identifier
                             "737472"
                             :types
                             ())
                         :is_op
                         false
                         :is_prefix_op
                         false)
                     (ConBind
                         :identifier
                         "C2AC"
                         :param
                         (TyconTypeExpr
                             :qualifiers
                             ()
                             :identifier
                             "666F726D756C61"
                             :types
                             ())
                         :is_op
                         true
                         :is_prefix_op
                         true)
                     (ConBind
                         :identifier
                         "E288A7"
                         :param
                         (TupleTypeExpr
                             :types
                             (
                                 (TyconTypeExpr
                                     :qualifiers
                                     ()
                                     :identifier
                                     "666F726D756C61"
                                     :types
                                     ())
                                 (TyconTypeExpr
                                     :qualifiers
                                     ()
                                     :identifier
                                     "666F726D756C61"
                                     :types
                                     ())
                             ))
                         :is_op
                         true
                         :is_prefix_op
                         false)
                     (ConBind
                         :identifier
                         "E288A8"
                         :param
                         (TupleTypeExpr
                             :types
                             (
                                 (TyconTypeExpr
                                     :qualifiers
                                     ()
                                     :identifier
                                     "666F726D756C61"
                                     :types
                                     ())
                                 (TyconTypeExpr
                                     :qualifiers
                                     ()
                                     :identifier
                                     "666F726D756C61"
                                     :types
                                     ())
                             ))
                         :is_op
                         true
                         :is_prefix_op
                         false)
                     (ConBind
                         :identifier
                         "E28692"
                         :param
                         (TupleTypeExpr
                             :types
                             (
                                 (TyconTypeExpr
                                     :qualifiers
                                     ()
                                     :identifier
                                     "666F726D756C61"
                                     :types
                                     ())
                                 (TyconTypeExpr
                                     :qualifiers
                                     ()
                                     :identifier
                                     "666F726D756C61"
                                     :types
                                     ())
                             ))
                         :is_op
                         true
                         :is_prefix_op
                         false)
                     (ConBind
                         :identifier
                         "E28AA5"
                         :param
                         nullptr
                         :is_op
                         true
                         :is_prefix_op
                         false)
                 )))))
0018 (DeclTopDecl
         :decl
         (DtypeDecl
             :bindings
             ((DtypeBind
                 :identifier
                 "70616972"
                 :types
                 (
                     "2761"
                     "2762"
                 )
                 :constructors
                 ((ConBind
                     :identifier
                     "50616972"
                     :param
                     (TupleTypeExpr
                         :types
                         (
                             (VarTypeExpr
                                 :id
                                 "2761")
                             (VarTypeExpr
                                 :id
                                 "2762")
                         ))
                     :is_op
                     false
                     :is_prefix_op
                     false))))))
0019 ERROR
0020 ERROR
0021 ERROR
0022 (DeclTopDecl
         :decl
         (DtypeDecl
             :bindings
             ((DtypeBind
                 :identifier
                 "70616972"
                 :types
                 ()
                 :constructors
                 ((ConBind
                     :identifier
                     "70616972"
                     :param
                     (TupleTypeExpr
                         :types
                         (
                             (TyconTypeExpr
                                 :qualifiers
                                 ()
                                 :identifier
                                 "696E74"
                                 :types
                                 ())
                             (TyconTypeExpr
                                 :qualifiers
                                 ()
                                 :identifier
                                 "737472696E67"
                                 :types
                                 ())
                         ))
                     :is_op
                     false
                     :is_prefix_op
                     false))))))
0024 ERROR
0025 ERROR
0026 ERROR
0027 ERROR
0028 ERROR
0029 ERROR
0030 ERROR
0031 ERROR
0032 ERROR
0033 ERROR
0034 ERROR
0035 ERROR
0036 ERROR
0037 ERROR
0038 ERROR
0039 ERROR
0040 (EndOfFileTopDecl)
